File: 2018/Day02/solution.js
const fs = require('fs');

// PART ONE
fs.readFile('./input.txt', 'utf8', (err, data) => {
  const arr = data.split('\r\n');

  // Check each entry in the array for letters that occur only 2 or 3 times
  // Count into object
  let checkedArray = arr.map(str => {
    let obj = {};
    str.split('').forEach(letter => {
      if (!obj[letter]) {
        obj[letter] = 0;
      }
      obj[letter]++;
    });
    return obj;
  });

  // Push entries with at least once exactly 2 letters into array twos
  // Entries with at least once exactly 3 letters into threes
  let two = [];
  let three = [];

  checkedArray.forEach(obj => {
    let temp = Object.values(obj);

    if (temp.includes(2)) {
      two.push(obj);
    }
    if (temp.includes(3)) {
      three.push(obj);
    }
  });

  // Multiply array lengths
  let result = two.length * three.length;

  // Log result
  console.log(result);
});

// PART TWO
fs.readFile('./input.txt', 'utf8', (err, data2) => {
  var arr2 = data2.split('\r\n');

  arr2.forEach(item => {
    let positionDiff = 0;

    arr2.forEach(check => {
      let countDiff = 0;
      for (let i = 0; i < item.length; i++) {
        if (item[i] !== check[i]) {
          positionDiff = i;
          countDiff++;
        }
      }
      if (countDiff === 1) {
        let result = item.split('');
        result.splice(positionDiff, 1);
        console.log('item: ', result.join(''));
      }
    });
  });
});

File: 2018/Day02/solution.py
from collections import Counter
import difflib

data = open("./input.txt").read().splitlines()

# Part 1
twos = []
threes = []

for str in data:
    if 2 in Counter(str).values():
        twos.append(str)
    if 3 in Counter(str).values():
        threes.append(str)

print(len(twos) * len(threes))

# PART 2
for i in range(0, len(data)):
    first_string = data[i]
    for j in range(i + 1, len(data)):
        second_string = data[j]
        diff = 0
        for idx in range(len(first_string)):
            if first_string[idx] != second_string[idx]:
                diff += 1
        if diff == 1:
            print(first_string)
            print(second_string)
